fdbserver/DDTxnProcessor.actor.cpp

```diff
@@ -0,0 +1,93 @@
+/*
+ * DDTxnProcessor.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbserver/DDTxnProcessor.h"
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+class DDTxnProcessorImpl {
+	friend class DDTxnProcessor;
+
+	// return {sourceServers, completeSources}
+	ACTOR static Future<IDDTxnProcessor::SourceServers> getSourceServersForRange(Database cx, KeyRangeRef keys) {
+		state std::set<UID> servers;
+		state std::vector<UID> completeSources;
+		state Transaction tr(cx);
+
+		loop {
+			servers.clear();
+			completeSources.clear();
+
+			tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
```

Comment:
Do we need set `FDBTransactionOptions::READ_SYSTEM_KEYS` here?

Reply:
It's a copy-paste from the old code, so...
I believe now even though without `FDBTransactionOptions::READ_SYSTEM_KEYS` we still can read system key. But in case of future we may have more strict check, I'll add it.